{% extends 'base.html' %}

{% block content %}
<div id="main_structure">

   {% include "nav.html" %}

   <div class="container main-container">
      <div class="tab-content" data-tabsetid="2132">
         <table id="tutorialTable">
            <tr valign="top">
               <td style="width: 200px">

                  <a>Overview of DoSurvive</a>
                  <hr>

                  <b>Browse Database</b>
                  <p></p>

                  <ul id="overviewSearch">

                     <li>
                        <a>Select a Cancer & a Gene</a>
                     </li>
                     <li>
                        <a>Select a Cancer</a>
                     </li>
                     <li>
                        <a>Select a Gene</a>
                     </li>
                     <li>
                        <a>*Explore methylation level and survival</a>
                     </li>
                  </ul>

                  <b>Explore Combined Effects</b>
                  <p></p>

                  <ul id="overviewInquiry">

                     <li>
                        <a>Two genes</a>
                     </li>
                     <li>
                        <a>More than Two Genes (>2 genes)</a>
                     </li>
                  </ul>

                  <b>Statistics Analysis</b>
                  <p></p>

                  <ul id="overviewBrowse">
                     <li>
                        <a id="log_rank_test">Log-Rank test</a>
                     </li>
                     <li>
                        <a>Cox regression</a>
                     </li>
                     <li>
                        <a>Accelerated Failure-Time model</a>
                     </li>
                     <li>
                        <a id="exp_methylation">Expression/methyaltion profiles</a>
                     </li>
                     <li>
                        <a>The correlation of mRNA & methylation</a>
                     </li>
                  </ul>

                  <b>About</b>
                  <p></p>

                  <ul id="overviewAbout">
                     <li>
                        <a>Download data</a>
                     </li>
                     <li>
                        <a href="http://dosurvive.lab.nycu.edu.tw/statistics">Statistics</a>
                     </li>
                     <li><a href="http://dosurvive.lab.nycu.edu.tw/faq">FAQ</a></li>
                  </ul>

               </td>
               <td>
                  <!--Overview of DoSurvive-->
                  <div class="tutorialPage">
                     <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                     DoSurvival is a web-based database for survival analysis, DoSurvive provides three
                        survival analysis method: the <span style="color:#0080FF;">log-rank test</span>, the <span style="color:#0080FF;">cox proportioanl hazard model</span> and the
                        <span style="color:#0080FF;">accelerated failure-time model</span>; and four survival types: overall survival, progression-free survival,
                        disease-free survival and disease-specific survival. For all three methods and four survival types,
                        DoSurvive offers five different quantiative features, namely expression level of protein, mRNA,
                        miRNA, lncRNA and the methylation levels of CpG sites, for up to 33 cancer types in TCGA.
                        With DoSurvive, users can:
                     <p></p>

                     <ol>
                        <li>Gene expression in cancers and survival analysis correlation</li>
            <li><span style="color:#0080FF;">Log-rank test</span> results of different molecular types in specific cancers</li>
            <li><span style="color:#0080FF;">Cox proportional hazard model</span> results of different molecular types in specific cancers</li>
            <li><span style="color:#0080FF;">Accelerated Failure-Time model (AFT)</span> results of different molecular types in specific cancers</li>
            <li>Gene expression level in normal and tumor samples</li>
            <li>CpG site & mRNA correlation in cancer</li>
            <li><span style="font-weight:bold; color:#0080FF;">Combined Effects, allow users to observe the relationship between multiple genes and survival, which is an innovative method by DoSurvival.</span></li>
                     </ol>

                     <div style="text-align: center;">
                        <img src="static/images/DoSurvive_dodo (1).png" style="width: 80%;" />
                     </div>

                     </div>
                     </div>

                  <!--Select cancer & gene-->
                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        Select a Cancer & a Gene</div>

                        <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                        A user who is interested in a specific cancer and a specific target feature can mouseover "Browse Database" and
                           then go to "Select Cancer & Gene" to browse the database.<br><br>
                           Step 1. Select a molecular type of interest from the list (red box shown below).
                           <br>
                           <div>
                              <img src="static/images/do1.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                              <!--<img src="static/images/tutorial_figure1.png" style="width: 80% border:2px #ccc solid;padding:5" />-->
                           </div>
                           <br>
                        </div>
                        <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                           Step 2. Choose the cancer type of interest from the dropdown list.<br>
                           *There will be different lists of cancers depending on the molecular type
                           chosen in Step 1.
                           <br>
                           <div>
                              <img src="static/images/2.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                        </div>

                     <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                           Step 3. Enter a gene name and click "Browse" to obtain the survival analysis results.
                           Both the original p-values and adjusted p-values (BH corrected) will be listed.
                           <br>
                           <div>
                              <img src="static/images/3.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 4. Choose a statistical method to explore the result. The user can click the
                           survival mode of interest and the corresponding plots or tables will be generated.
                           For example, if the user selects "Cox" and clicks "Progression-Free" (blue
                           box) a hazard ratio plot and a table of  detailed numbers will be shown (red box). If
                           the user selects "Log Rank", an interactive Kaplan-Meier plot will be shown (red box).
                           *The user can customize the color/thickness of the lines in the Kaplan-Meier plot. For
                           more detail, see <a class="log_rank_test">Log-Rank test</a>.
                           <div>
                              <img src="static/images/4.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 5. (optional) The user may remove feature(s) from the Cox regression or AFT model by unchecking the feature(s) and
                           clicking "calculate". For example, after unchecking "Sex" and clicking "calculate", the results from the new Cox regression
                           model are calculated on the fly, with output as follows:
                           <div>
                              <img src="static/images/5.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 6. (optional) The user can also specify a criteria for patients to be included in the analysis.
                           By clicking "Download clinical data", a csv file for the clinical information will be downloaded. The user
                           can select a subset of patients and upload the file for reanalysis. The reanalyzed results will be shown below the original
                           results for easy comparison.
                           <div>
                              <img src="static/images/6.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 7. (optional) For mRNA, the user can further explore the correlation between a specific mRNA gene and its upstream CpG
                           sites. By clicking the button labeled 450k or 27k, a pop-up window will show all the upstream CpG sites for
                        the specified mRNA. For more details, see <a class="exp_methylation">mRNA & methylation</a>.
                           <div>
                              <img src="static/images/7.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                        </div>
                     </div>

                  <!--Select cancer-->
                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        Select a Cancer</div>

                        <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                           A user who is interested in a specific cancer can mouseover "Browse Database" and then
                           go to "Select Cancer" to browse the database.
                           <br>
                           <br>
                           Step 1. Select a molecular type of interest from the list (red box shown below).
                           <div>
                              <img src="static/images/8.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                        </div>
                        <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                           Step 2. Choose the cacner type of interest from the dropdown list and click "Browse".
                           <br>
                           *There will be different lists of cancers depending on the molecular type chosen in Step 1.
                           <div>
                              <img src="static/images/9.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 3. Choose a statistical method to explore the results. All the features for the selected
                           molecular type for the selected cancer will be shown and ranked by their adjusted p-value (BH corrected)
                           in a table. Significant risks and protective features will be shown in red and green, respectively.
                           <div>
                              <img src="static/images/10.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 4. The user can click the feature (gene) of interest and the corresponding plot or table
                           will be generated below the table. For example, if the user selects "Cox" and clicks a specific gene
                           (blue box), a hazard ratio plot and a table with detailed numbers will be shown (red box). If the user
                           selects "Log rank", an interactive Kaplan-Meier plot will be shown.<br>
                           *The user can customize for color/thickness of the lines in the Kaplan-Meier plot. For more details,
                           see <a class="log_rank_test">Log-rank test</a>.
                           <div>
                              <img src="static/images/11.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 5. The user can select a survival type of interest to explore (red box) and the corresponding
                           table will be shown on the fly.
                           <div>
                              <img src="static/images/12.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 6. (optional) The user may remove feature(s) from the Cox regression or AFT model by unchecking
                           the features and clicking "calculate", the results from the new Cox regression model is calculated on
                           the fly, as follows:
                           <div>
                              <img src="static/images/13.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 7. (optional) The user can also specify the criteria for the patients to be included in the
                           analysis. By clicking "Download clinical data", a csv file for the clinical information will be
                           downloaded. The user can select a subset of patients and upload the file for analysis. The reanalyzed
                           result will be shown below the original results for easy comparison.
                           <div>
                              <img src="static/images/14.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 8. (optional) For mRNA, the user can further explore the correlation between a specific mRNA gene and its upstream CpG
                           sites. By clicking the button labeled 450k or 27k, a pop-up window will show all the upstream CpG sites for
                           the specified mRNA. For more details, see <a class="exp_methylation">mRNA & methylation</a>.
                           <div>
                              <img src="static/images/15.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           <div>
                              <img src="static/images/16.PNG" style="width: 80%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           </div>
                  </div>

                  <!--Select gene-->
                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        Select a Gene</div>


                        <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                           The user who is interested in specific gene can mouseover "Browse Database" and then go to
                           "Select Gene" to browse the database.
                           <br>
                           <br>
                           Step 1. Select "Select Gene" (red box below).
                           <div>
                              <img src="static/images/17.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 2 . Choose the cancer type of interest form the dropdown list and click "Browse".
                           <div>
                              <img src="static/images/18.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 3. Choose a Statistical method from the dropdown list to explore the results. The survival
                           analysis results for the queried gene in all the cancer types will be shown and ranked by their
                           adjusted p-values (BH corrected) in a table. Significant risks and protective features will be shown
                           in red and green, respectively.
                           <div>
                              <img src="static/images/19.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 4. The user can click the feature (gene) of interest and the corresponding plot or table will
                           be generated below the table. For example, if the user selects "Log Rank" and clicks Mesothelioma (MESO)
                           (blue box) an interactive Kaplan-Meier plot will be shown (red box).<br>
                           *The user can customize the color/thickness of lines in Kaplan-Meier plot. For more details, see
                           <a class="log_rank_test">Log-rank test</a>.
                           <div>
                              <img src="static/images/20.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 5. The user can select a survival type of interest to explore (red box) and the corresponding table will
                           be shown on the fly.
                           <div>
                              <img src="static/images/21.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 6. (optional) The user may remove feature(s) from the Cox regression or AFT model by unchecking the feature(s)
                           and clicking "calculate" (blue box). In the following example, after unchecking "Stage" and clicking "calculate" (which
                           violates PH assumption), the results from the new Cox regression model are calculated on the fly.
                           <div>
                              <img src="static/images/22.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 7. (optional) The user can also specify the criteria for patients to include in the analysis. By clicking
                           "Download clinical data", a csv file for the clinical information will be downloaded. The user can select a subset
                           of patients and upload the file for reanalysis. The reanalyzed result will be shown below the original results
                           for easy comparison.
                           <div>
                              <img src="static/images/23.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 8. (optional) For mRNA, the user can further explore the correlation between a specific mRNA gene and its upstream CpG
                           sites. By clicking the button labeled 450k or 27k, a pop-up window will show all the upstream CpG sites for
                           the specified mRNA. For more details, see <a class="exp_methylation">mRNA & methylation</a>.
                           <div>
                              <img src="static/images/24.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                        </div>
                  </div>


                  <!--Explore methylation & survival-->
                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        Explore Methylation Level & Survival</div>


                        <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                           The user who is interested in the methylation levels of CpG sites can explore the survival analysis
                           results for CpG sites by querying their downstream genes.
                           <br>
                           <br>
                           Step 1. Go to the table of the survival analysis results for the gene of interest and
                           click button labeled 450k or 27k. A pop-up window will show the survival analysis results
                           for all the upstream CpG sites for the gene of interest.
                           <div>
                              <img src="static/images/25.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 2. Choose a statistical method from the dropdown list to explore the results.
                           The survival analysis results for the queried gene in all the cancer types will be shown
                           and ranked by their adjusted p-valuee (BH corrected) in a table. Significant risks
                           and protective features will be shown in red and green, respectively.
                           <div>
                              <img src="static/images/26.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 3. The user can click the CpG site of interest and the corresponding plot or table will
                           be generated below the table. For example, if the user selects "Log Rank" and selects a CpG
                           site (blue box), an interactive Kaplan-Meier plot will be shown (red box).<br>
                           *The user can customize the color/thickness of the lines in the Kaplan-Meier plot. For more
                           details, see <a class="log_rank_test">Log-Rank test</a>.
                           <div>
                              <img src="static/images/27.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Stpe 4. The user can select a survival type of interest (red box) and the corresponding
                           table will be shown on the fly.
                           <div>
                              <img src="static/images/28.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 5. (optional) The user may remove feature(s) from the Cox regression or AFT model by
                           unchecking the feature(s) and clicking "calculate", which violates the PH assumption, the results
                           from the new Cox regression model are calculated on the fly.
                           <div>
                              <img src="static/images/29.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                           Step 6. (optional) The user can also specify a criteria for patients to be included in the
                           analysis. By clicking "Download clinical data", a csv file for the clinical information will
                           be downloaded. The user can select a subset of patients and upload the file for reanalysis.
                           The reanalyzed results will be shown below the original results easy comparison.
                           <div>
                              <img src="static/images/30.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                           </div>
                           <br>
                        </div>
                        </div>


                  <!--Two genes-->
                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        Two Genes</div>

                     <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                        A user who is interested in the combined effect of two genes in survival, can mouseover "Explore
                        Combined Effects" and then go to "Two genes" to analyze the combined effect with the log rank test, the
                        cox proportional hazard or the accelerated failure-time (AFT) model.
                        <br>
                        <br>
                        Step 1. Select a combination of two genes from the list (red box shown below).
                        <div>
                           <img src="static/images/31.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        Step 2. Choose the cancer type of interest from the dropdown list.
                        <div>
                           <img src="static/images/32.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        Step 3. Enter two genes in the two text boxes (blue box) and select the groups (green box)
                        that are to be compared using log-rank test. Click "calculate".
                        <div>
                           <img src="static/images/33.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        Step 4. Choose a statistical method to explore the results. The user click the survival type of
                        interest and the corresponding plot or table will be generated. For example, if the user selects
                        "Log Rank" and clicks "Progression-Free" (blue box), an interactive Kaplan-Meier will be shown (red box).
                        If the user selects "Cox" and click "Progression-Free", a hazard ratio plot and a table with detailed
                        numbers will be shown.<br>
                        *The user can customize the color/thickness of the lines in the Kaplan-Meier plot. For more details,
                        see <a class="log_rank_test">Log-Rank test</a>.
                        <div>
                           <img src="static/images/34.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        Step 5. (optional) The user may remove feature(s) from the Cox regression or AFT model by unchecking
                        the feature(s) and clicking "calculate". For example, after unchecking "Sex" and clicking "calculate",
                        the results from the new Cox regression model are calculated on the fly.
                        <div>
                           <img src="static/images/35.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        Step 6. (optional)  The user can also specify the criteria for patients to be included in the
                        analysis. By clicking "Download clinical data", a csv file with clinical information  will be downloaded.
                        The user can select a subset of patients and upload the file for reanalysis. The reanalyzed results will be
                        shown below the original results for easy comparison.
                        <div>
                           <img src="static/images/36.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                     </div>
                  </div>


                  <!-- > two genes-->
                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        More than Two Genes (>2 Genes)</div>

                     <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                        A user who is interested in the combined effect of more than two genes on survival can mouseover
                        "Explore Combined Effects" and then click ">2 genes" to analyze the combined effect with
                        the Cox proportional hazards model and the accelerated failure-time (AFT) model.
                        <br>
                        <br>
                        Step 1. Select ">2 genes" (red box shown below).
                        <div>
                              <img src="static/images/37.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        Step 2. Choose the cancer type of interest from the dropdown list.
                        <div>
                              <img src="static/images/38.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        Step 3. Enter the target mRNA genes, miRNA and lncRNA, in the first, second and third text box,
                        respectively. The entered gene will be listed directly below the text box and the user can remove
                        the entered gene by clicking wanted features. After all the genes have been filled in, click "calculate"
                        to implement the analysis.
                        <div>
                              <img src="static/images/39.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        Step 4. Choose a statistical method for analysis. The user can click the survival
                        model of interest and the corresponding plot or table will be generated. For example, if the user
                        selects "Cox" and clicks "Progression-Free" (blue box) a hazard ratio plot and a table with
                        detailed numbers will be shown (red box).
                        <div>
                              <img src="static/images/40.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        Step 5. (optional) The user may remove feature(s) from the Cox regression or AFT model by
                        unchecking the feature(s) and clicking "calculate". For example, after unchecking "Sex" and
                        clicking "calculate", the results from the new Cox regression model are calculated on the fly.
                        <div>
                              <img src="static/images/41.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        Step 6. (optional) A user can also specify the criteria for patients to be included in the analysis.
                        By clicking "Download clinical data", a csv file for the clinical information will be downloaded.
                        The user can select a subset of patients and upload the file for reanalysis. The reanalyzed results
                        will be shown below the original results for easy comparison.
                        <div>
                              <img src="static/images/42.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                     </div>
                  </div>



                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        Log-rank Test</div>
                     <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                        DoSurvive uses the <span style="color: red;">median to divide the data into two
                           groups</span>, a high-performance group and a low-performance group, DoSurvive then
                        conducts a Log-rank test. The cut-off is the expression is less than half is 0 or
                        NA. The log-rank test yield the overall, disease-free, disease-specific, and progression-free
                        p-value as well as the BH adjusted p-values (False-discovery
                        rates).
                        <div>
                           <img src="static/images/43.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        The user can click a gene they are interested in, and this will result in overall survival, disease-free
                        survival, disease-specific survival and progression-free survival Kaplan-Meier plots. By default, the red
                        line represents the high-performance group, the blue line represents the low-performance group, and n
                        represents the sample sizes. DoSurvive allows the user to customize the Kaplan-Meier plot, i.e line thickness
                        and color, follow up threshold (years), and the percentages used to separate two groups of patients in Kapan-Merier plot. Clicking "Submit" the Kaplan-Meier plot will be reproduced (red box).
                        <div>
                           <img src="static/images/44.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                     </div>
                  </div>


                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        Cox Proportional Hazard model (Cox Regression)</div>

                     <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                        The purpose of the Cox model is to evaluate the simultaneous effect of several
                        factors on survival. In other words, it allows us to examine how specific
                        factors influence the rate of a particular event happening (e.g., infection,
                        death) at a particular point in time. This rate is commonly referred to as the
                        hazard rate. Predictor variables (or factors) are usually termed covariates in
                        the survival-analysis literature.
                        <p></p>
                        The Cox model is expressed in terms of the hazard function, denoted by h(t), which briefly can
                        be interpreted as the relative risk of dying between times t and t+s, given survival to time t
                        (where s is a small value). In the cox model, the form of the hazard function is given by :
                        <div>
                           <img src="static/images/45.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        In the Cox regression results, a p-value of < 0.05 for coefficient βi indicates that gene expression i
                        has a significant impact on the hazard (i = 1,2,....,p). Red signifies a p-value of < 0.05 and a positive
                        coefficient (βi > 0), indicating that the increase in the expression of this gene has a negative impact on
                        survival; Green signifies a p-value of < 0.05 and a negative coefficient (βi < 0), indicating that the
                        increase in the expression of this gene has a positive impact on survival.
                        <div>
                           <img src="static/images/46.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        In "Cox regression results", the user can also click a gene/survival type to plot the hazard ratio
                        plot. DoSurvive provides both a hazard ratio plot and a table for the hazard ratios, confidence intervals,
                        p-values, and proportional hazard assumption p-values. The user can remove unwanted features from the
                        Cox model and reanalyze by unchecking variables (blue box) to get new results (red box).
                        <div>
                           <img src="static/images/47.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        The user can also specify the criteria for patients to be included in the analysis. By clicking
                        "Download clinical data", a csv file for the clinical information will be downloaded. The user
                        can select a subset of patients and upload the file for reanalysis. The reanalyzed results will be
                        shown below the original results for easy comparison.
                        <div>
                           <img src="static/images/48.png" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                     </div>
                  </div>

                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        Accelerated Failure-Time model (AFT)</div>

                     <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                     In survival analysis, an accelerated failure time model (AFT
                     model) is a parametric model that provides an alternative to the commonly used
                     proportional hazards models. Whereas a proportional hazards model assumes that the
                     effect of a covariate is to multiply the hazard by some constant, an AFT model
                     assumes that the effect of a covariate is to <span style="color: red;">accelerate</span> or <span
                        style="color: red;">decelerate</span> the life course of a disease by some
                     constant. <span style="color: red;">Red indicates p-value < 0.05 and coefficient is negative</span>
                           , indicates that the increase in the expression of this gene
                           have a <span style="color: red;">negative impact</span> on survival which ;
                           <span style="color: #00EC00;">Green indicates p-value < 0.05 and coefficient is positive</span>,
                                 indicates that the increase in the expression of this
                                 gene has a <span style="color: #00EC00;">positive impact</span> on
                                 survival. In full generality, the accelerated failure time model can be
                                 specified as :
                                 <div style="text-align:center;">
                                    <img src="static/images/49.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                                 </div>
                                 <br>


                                 In the AFT results, the user can also click a gene/survival type to plot the time ratio.
                                 <div>
                                    <img src="static/images/50.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                                 </div>
                                 <br>
                                 The user can remove unwanted features from the AFT model and reanalyzed by unchecking
                                 variables (blue box) to get new results (red box).
                                 <div>
                                    <img src="static/images/51.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                                 </div>
                                 <br>
                                 The user can also specify the criteria for patients to be included in the analysis. By
                                 clicking "Download clinical data", a csv file for the clinical information will be downloaded.
                                 The user can select a subset of patients and upload the file for reanalysis. The reanalyzed
                                 results will be shown below the original result for easy comparison.
                                 <div>
                                    <img src="static/images/52.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                                 </div>
                     </div>
                     </div>

                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        Expression/Methylation Profiles</div>

                     <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                        DoSurvive provides the profiles for the expression levels of a specific gene
                        (mRNA/miRNA/lncRNA) or protein and methylation levels for CpG sites in different
                        cancer types.
                        <br>
                        <br>
                        * Gene/protein: In the "Expression" tab, Expression (Normal) represents the average
                        expression of all normal samples; Expression (Tumor) represents the average expresssion
                        of all tumor samples. DoSurvive also calculates the Wilcoxon-rank sum test to confirm
                        whether there is a significant difference between the expression levels in normal samples
                        and in tumor samples.
                        <div>
                           <img src="static/images/53.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        By clicking on features of interest, an interactive box plot of the expression levels
                        across cancer types will be generated.
                        <div>
                           <img src="static/images/54.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                        <br>
                        * CpG sites: For a gene of interest, the user can retrieve the methylation levels of its upstream
                        CpG sites via click buttons labeled as 27k or 450k in the table (blue box). A pop-up window for the
                        survival analysis results and Expression profile of the methylation levels from the upstream CpG
                        sites will be shown (red box).
                        <div>
                           <img src="static/images/55.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                        </div>
                     </div>
                  </div>

                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        The Correlation of mRNA & Methylation</div>
                     <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                     If a user wants to further explore how methylation regulates mRNA, DoSurvive provides
                        the relationship between the expression level of mRNA and the methylation levels of
                        its upstream CpG sites.
                        <br>
                        <br>
                        In the table for mRNA, the user can click 27k or 450k (blue box) to display the
                        analysis results (red box) for the methylation levels detected from different methylation
                        arrays, i.e. Illumina Infinium HumanMethylation27 BeadChip and Illumina Infinium HumanMethylation450 BeadChip.
                     <div>
                        <img src="static/images/56.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                     </div>
                     <br>
                        By clicking the "Expression" tab, the user can select the CpG site of interest and then
                        click "Start" to explore the correlation between the expression level of the selected gene
                        and the methylation level of its upstream CpG site(s).
                     </div>
                     <div>
                        <img src="static/images/57.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                     </div>
                  </div>

                  <div class="tutorialPage" style="display: none;">
                     <div style="color: red; font-size: 20px; font-weight: bold; padding-bottom: 5px;">
                        Download data</div>

                     <div style="color: black; font-size: 15px;  padding-bottom: 2px;">
                     Tables of all analysis results can be downloaded. The user can click "Download", "Copy", "CSV", "Excel", "PDF" or "Print"
                        to get the whole table.
                     <div>
                        <img src="static/images/download1.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                     </div>
                     <br>
                     <div>
                        <img src="static/images/download2.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                     </div>
                     <br>
                     In addition to the results table, all the clinical data used in the survival analysis can be downloaded
                        for further analysis.
                     <div>
                        <img src="static/images/download3.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                     </div><br>
                        The downloaded clinical data has been processed, "pathologic_stage" has been classified as a categorical variable, "0" to represent
                        early stage, "1" to represent lately stage. "gender" has been classified as a categorical variable, "0" to represent female, "1" to
                        represent male. Expression data has been normalized (red box). The remaining columns provide user reference and do not participate in the regression model.
                     <div>
                        <img src="static/images/download4.PNG" style="width: 100%; border:2px #ccc solid;padding:5px;"><br>
                     </div>
                  </div>
                  </div>
<!--
<div class="tutorialPage" style="display: none;">
      <div class="tab-content" data-tabsetid="2132">
         <h2>Statistics on the cancer types and number of samples in <font color="#02DF82">DoSurvive</font>.</h2>
         <br>

         <table class="table table-bordered">
            <tbody>
               <tr style="text-align: center;">
                  <td style="text-align: left; vertical-align:middle;" rowspan="2">
                     Cancer Types
                  </td>
                  <td colspan="2">protein</td>
                  <td colspan="2">mRNA</td>
                  <td colspan="2">miRNA</td>
                  <td colspan="2">LncRNA</td>
                  <td colspan="2">methylation27k</td>
                  <td colspan="2">methylation450k</td>
               </tr>
               <tr style="text-align: center;">
                  <td>Tumor</td>
                  <td>Normal</td>
                  <td>Tumor</td>
                  <td>Normal</td>
                  <td>Tumor</td>
                  <td>Normal</td>
                  <td>Tumor</td>
                  <td>Normal</td>
                  <td>Tumor</td>
                  <td>Normal</td>
                  <td>Tumor</td>
                  <td>Normal</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Adrenocortical carcinoma(ACC)</td>
                  <td>46</td>
                  <td>-</td>
                  <td>79</td>
                  <td>-</td>
                  <td>80</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>80</td>
                  <td>-</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Bladder Urothelial Carcinoma(BLCA)</td>
                  <td>344</td>
                  <td>-</td>
                  <td>408</td>
                  <td>19</td>
                  <td>409</td>
                  <td>19</td>
                  <td>252</td>
                  <td>19</td>
                  <td>-</td>
                  <td>-</td>
                  <td>412</td>
                  <td>21</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Breast invasive carcinoma(BRCA)</td>
                  <td>887</td>
                  <td>45</td>
                  <td>1093</td>
                  <td>112</td>
                  <td>1078</td>
                  <td>104</td>
                  <td>837</td>
                  <td>105</td>
                  <td>314</td>
                  <td>27</td>
                  <td>784</td>
                  <td>98</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Cervical squamous cell carcinoma and endocervical adenocarcinoma(CESC)
                  </td>
                  <td>173</td>
                  <td>-</td>
                  <td>304</td>
                  <td>3</td>
                  <td>307</td>
                  <td>3</td>
                  <td>196</td>
                  <td>3</td>
                  <td>-</td>
                  <td>-</td>
                  <td>307</td>
                  <td>3</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Cholangiocarcinoma(CHOL)</td>
                  <td>30</td>
                  <td>-</td>
                  <td>36</td>
                  <td>9</td>
                  <td>36</td>
                  <td>9</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>36</td>
                  <td>9</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Colon adenocarcinoma(COAD)</td>
                  <td>360</td>
                  <td>-</td>
                  <td>285</td>
                  <td>41</td>
                  <td>406</td>
                  <td>-</td>
                  <td>157</td>
                  <td>-</td>
                  <td>165</td>
                  <td>37</td>
                  <td>295</td>
                  <td>38</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Lymphoid Neoplasm Diffuse Large B-cell Lymphoma(DLBC)</td>
                  <td>33</td>
                  <td>-</td>
                  <td>48</td>
                  <td>-</td>
                  <td>47</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>48</td>
                  <td>-</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Esophageal carcinoma(ESCA)</td>
                  <td>126</td>
                  <td>-</td>
                  <td>184</td>
                  <td>11</td>
                  <td>184</td>
                  <td>13</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>185</td>
                  <td>16</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Glioblastoma multiforme(GBM)</td>
                  <td>238</td>
                  <td>-</td>
                  <td>160</td>
                  <td>5</td>
                  <td>-</td>
                  <td>5</td>
                  <td>154</td>
                  <td>-</td>
                  <td>285</td>
                  <td>-</td>
                  <td>141</td>
                  <td>2</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Head and Neck squamous cell carcinoma(HNSC)</td>
                  <td>357</td>
                  <td>-</td>
                  <td>520</td>
                  <td>44</td>
                  <td>523</td>
                  <td>44</td>
                  <td>426</td>
                  <td>42</td>
                  <td>-</td>
                  <td>-</td>
                  <td>528</td>
                  <td>50</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Kidney Chromophobe(KICH)</td>
                  <td>63</td>
                  <td>-</td>
                  <td>66</td>
                  <td>25</td>
                  <td>66</td>
                  <td>25</td>
                  <td>66</td>
                  <td>25</td>
                  <td>-</td>
                  <td>-</td>
                  <td>66</td>
                  <td>-</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Kidney renal clear cell carcinoma(KIRC)</td>
                  <td>478</td>
                  <td>-</td>
                  <td>533</td>
                  <td>72</td>
                  <td>516</td>
                  <td>71</td>
                  <td>448</td>
                  <td>67</td>
                  <td>219</td>
                  <td>199</td>
                  <td>319</td>
                  <td>160</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Kidney renal papillary cell carcinoma(KIRP)</td>
                  <td>215</td>
                  <td>-</td>
                  <td>290</td>
                  <td>32</td>
                  <td>291</td>
                  <td>34</td>
                  <td>198</td>
                  <td>30</td>
                  <td>16</td>
                  <td>5</td>
                  <td>275</td>
                  <td>45</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Acute Myeloid Leukemia(LAML)</td>
                  <td>-</td>
                  <td>-</td>
                  <td>173</td>
                  <td>-</td>
                  <td>188</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>194</td>
                  <td>-</td>
                  <td>194</td>
                  <td>-</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Brain Lower Grade Glioma(LGG)</td>
                  <td>430</td>
                  <td>-</td>
                  <td>516</td>
                  <td>-</td>
                  <td>512</td>
                  <td>-</td>
                  <td>486</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>516</td>
                  <td>-</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Liver hepatocellular carcinoma(LIHC)</td>
                  <td>184</td>
                  <td>-</td>
                  <td>371</td>
                  <td>50</td>
                  <td>372</td>
                  <td>50</td>
                  <td>200</td>
                  <td>50</td>
                  <td>-</td>
                  <td>-</td>
                  <td>377</td>
                  <td>50</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Lung adenocarcinoma(LUAD)</td>
                  <td>365</td>
                  <td>-</td>
                  <td>515</td>
                  <td>59</td>
                  <td>513</td>
                  <td>46</td>
                  <td>488</td>
                  <td>58</td>
                  <td>126</td>
                  <td>24</td>
                  <td>458</td>
                  <td>32</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Lung squamous cell carcinoma(LUSC)</td>
                  <td>328</td>
                  <td>-</td>
                  <td>501</td>
                  <td>51</td>
                  <td>478</td>
                  <td>45</td>
                  <td>220</td>
                  <td>17</td>
                  <td>133</td>
                  <td>27</td>
                  <td>370</td>
                  <td>42</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Mesothelioma(MESO)</td>
                  <td>63</td>
                  <td>-</td>
                  <td>87</td>
                  <td>-</td>
                  <td>87</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>87</td>
                  <td>-</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Ovarian serous cystadenocarcinoma(OV)</td>
                  <td>426</td>
                  <td>-</td>
                  <td>304</td>
                  <td>-</td>
                  <td>453</td>
                  <td>-</td>
                  <td>412</td>
                  <td>-</td>
                  <td>584</td>
                  <td>12</td>
                  <td>10</td>
                  <td>-</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Pancreatic adenocarcinoma(PAAD)</td>
                  <td>123</td>
                  <td>-</td>
                  <td>178</td>
                  <td>4</td>
                  <td>178</td>
                  <td>4</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>184</td>
                  <td>10</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Pheochromocytoma and Paraganglioma(PCPG)</td>
                  <td>80</td>
                  <td>-</td>
                  <td>179</td>
                  <td>3</td>
                  <td>179</td>
                  <td>3</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>179</td>
                  <td>3</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Prostate adenocarcinoma(PRAD)</td>
                  <td>352</td>
                  <td>-</td>
                  <td>497</td>
                  <td>52</td>
                  <td>494</td>
                  <td>52</td>
                  <td>374</td>
                  <td>52</td>
                  <td>-</td>
                  <td>-</td>
                  <td>498</td>
                  <td>50</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Rectum adenocarcinoma(READ)</td>
                  <td>131</td>
                  <td>-</td>
                  <td>94</td>
                  <td>10</td>
                  <td>143</td>
                  <td>-</td>
                  <td>71</td>
                  <td>-</td>
                  <td>68</td>
                  <td>5</td>
                  <td>98</td>
                  <td>7</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Sarcoma(SARC)</td>
                  <td>223</td>
                  <td>-</td>
                  <td>259</td>
                  <td>2</td>
                  <td>259</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>261</td>
                  <td>4</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Skin Cutaneous Melanoma(SKCM)</td>
                  <td>353</td>
                  <td>-</td>
                  <td>469</td>
                  <td>1</td>
                  <td>448</td>
                  <td>2</td>
                  <td>226</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>470</td>
                  <td>2</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Stomach adenocarcinoma(STAD)</td>
                  <td>357</td>
                  <td>-</td>
                  <td>415</td>
                  <td>35</td>
                  <td>436</td>
                  <td>41</td>
                  <td>285</td>
                  <td>33</td>
                  <td>48</td>
                  <td>25</td>
                  <td>395</td>
                  <td>2</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Testicular Germ Cell Tumors(TGCT)</td>
                  <td>118</td>
                  <td>-</td>
                  <td>150</td>
                  <td>-</td>
                  <td>150</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>150</td>
                  <td>-</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Thyroid carcinoma(THCA)</td>
                  <td>368</td>
                  <td>-</td>
                  <td>501</td>
                  <td>59</td>
                  <td>502</td>
                  <td>59</td>
                  <td>497</td>
                  <td>59</td>
                  <td>-</td>
                  <td>-</td>
                  <td>503</td>
                  <td>56</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Thymoma(THYM)</td>
                  <td>90</td>
                  <td>-</td>
                  <td>120</td>
                  <td>2</td>
                  <td>124</td>
                  <td>2</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>124</td>
                  <td>2</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Uterine Corpus Endometrial Carcinoma(UCEC)</td>
                  <td>440</td>
                  <td>-</td>
                  <td>177</td>
                  <td>24</td>
                  <td>538</td>
                  <td>33</td>
                  <td>316</td>
                  <td>4</td>
                  <td>117</td>
                  <td>1</td>
                  <td>431</td>
                  <td>46</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Uterine Carcinosarcoma(UCS)</td>
                  <td>48</td>
                  <td>-</td>
                  <td>57</td>
                  <td>-</td>
                  <td>56</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>57</td>
                  <td>-</td>
               </tr>
               <tr style="text-align: center;">
                  <td style="text-align: left;">Uveal Melanoma(UVM)</td>
                  <td>12</td>
                  <td>-</td>
                  <td>80</td>
                  <td>-</td>
                  <td>80</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>-</td>
                  <td>80</td>
                  <td>-</td>
               </tr>
            </tbody>
         </table>
      </div>
</div>-->



<div class="tutorialPage" style="display: none;">


</div>

<script>
   // click tutorial side menu
   $('#tutorialTable a').each(function (index) {
      $(this).on('click', function () {
         $('.tutorialPage').hide();
         $('.tutorialPage').eq(index).show();
      })
   })

   $('.log_rank_test').on('click', function () {
         $('#log_rank_test').click();
      })

   $('.exp_methylation').on('click', function () {
         $('#exp_methylation').click();
      })

</script>
{% endblock %}
